plugins/amazonq/codetransform/jetbrains-community/src/software/aws/toolkits/jetbrains/services/codemodernizer/CodeModernizerSession.kt

```diff
@@ -394,17 +395,24 @@ class CodeModernizerSession(
             throw AlreadyDisposedException("Disposed when about to upload project artifact to s3")
         }
         val uploadStartTime = Instant.now()
-        try {
-            clientAdaptor.uploadArtifactToS3(
-                createUploadUrlResponse.uploadUrl(),
-                payload,
-                sha256checksum,
-                createUploadUrlResponse.kmsKeyArn().orEmpty(),
-            ) { shouldStop.get() }
-        } catch (e: Exception) {
-            LOG.error { "Unexpected error when uploading project artifact to S3: $e" }
-            throw e // pass along error to callee
+        for (i in 0..2) {
+            try {
+                clientAdaptor.uploadArtifactToS3(
+                    createUploadUrlResponse.uploadUrl(),
+                    payload,
+                    sha256checksum,
+                    createUploadUrlResponse.kmsKeyArn().orEmpty(),
+                ) { shouldStop.get() }
+                break
+            } catch (e: Exception) {
+                LOG.error { "Unexpected error when uploading project artifact to S3 on attempt ${i+1}/3: ${e.localizedMessage}" }
+                if (i == 2) {
+                    throw e
+                }
+                Thread.sleep((1000 * 2.0.pow(i)).toLong())
+            }
         }
+        LOG.info { "Upload to S3 succeeded" }
```

Comment:
I'm not a big fan of implementing our own retry logic rather than doing it at the SDK layer, but the existing S3 client's upload functionality requires a bucket name as a part of the request, and 1) I don't think we should have the internal S3 bucket names we use floating around this repo, 2) we use multiple different buckets, and 3) we don't have access to the bucket name in the response of our CreateUploadUrl — this API would have to be changed and then we can use the existing S3 client.

Reply:
Update: use `waitUntil` instead of a for loop